Tradução das Variáveis:
<ul>
<li>SSN: Número do Seguro Social</li>
<li>Occupation: Ocupação/profissão/cargo</li>
<li>Annual_Income: Renda anual</li>
<li>Monthly_Inhand_Salary: Salário base mensal</li>
<li>Num_Bank_Accounts: Número de contas no banco</li>
<li>Num_Credit_Card: Número de cartões de crédito</li>
<li>Interest_Rate: Taxa de juros do cartão de crédito</li>
<li>Num_of_Loan: Número de empréstimos/créditos</li>
<li>Type_of_Loan: Tipo de empréstimo/crédito</li>
<li>Delay_from_due_date: Atraso a partir da data de vencimento</li>
<li>Num_of_Delayed_Payment: Número de pagamentos atrasados</li>
<li>Changed_Credit_Limit: Limite de crétido alterado</li>
<li>Num_Credit_Inquiries: Número de consultas ao crédito</li>
<li>Credit_Mix:Mix de crédito</li>
<li>Outstanding_Debt: Dívida restante a ser paga (em USD)</li>
<li>Credit_Utilization_Ratio:A relação entre o crédito utilizado e o crédito total disponível</li>
<li>Credit_History_Age: Quanto tempo a pessoa tem o crédito</li>
<li>Payment_of_Min_Amount: Representa se somento foi pago o valor mínimo</li>
<li>Total_EMI_per_month: EMI se refere a um esquema de pagamento onde você pode converter suas despesas de cartão de crédito em parcelas mensais fáceis. Se aprovado, o valor é depositado na conta corrente ou poupança e as parcelas são lançadas nas faturas do cartão</li>
<li>Amount_invested_monthly: Quantia investida mensalmente</li>
<li>Payment_Behaviour: Se refere ao comportamento de pagamento do titular do cartão, incluindo a frequência e pontualidade dos pagamentos, o uso do crédito disponível e a gestão geral da dívida do cartão de crédito</li>
<li>Monthly_Balance: Saldo mensal</li>
<li>Credit_Score: Score</li>
</ul>

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('train.csv')
df.head()

C:\Users\marce\AppData\Local\Temp\ipykernel_16608\189615863.py:1: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('train.csv')


ID Customer_ID     Month           Name   Age          SSN Occupation  \
0  0x1602   CUS_0xd40   January  Aaron Maashoh    23  821-00-0265  Scientist   
1  0x1603   CUS_0xd40  February  Aaron Maashoh    23  821-00-0265  Scientist   
2  0x1604   CUS_0xd40     March  Aaron Maashoh  -500  821-00-0265  Scientist   
3  0x1605   CUS_0xd40     April  Aaron Maashoh    23  821-00-0265  Scientist   
4  0x1606   CUS_0xd40       May  Aaron Maashoh    23  821-00-0265  Scientist   

  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  ...  Credit_Mix  \
0      19114.12            1824.843333                  3  ...           _   
1      19114.12                    NaN                  3  ...        Good   
2      19114.12                    NaN                  3  ...        Good   
3      19114.12                    NaN                  3  ...        Good   
4      19114.12            1824.843333                  3  ...        Good   

   Outstanding_Debt Credit_Utilization_Ratio     Credit_History_Age  \
0            809.98                26.822620  22 Years and 1 Months   
1            809.98                31.944960                    NaN   
2            809.98                28.609352  22 Years and 3 Months   
3            809.98                31.377862  22 Years and 4 Months   
4            809.98                24.797347  22 Years and 5 Months   

   Payment_of_Min_Amount Total_EMI_per_month Amount_invested_monthly  \
0                     No           49.574949       80.41529543900253   
1                     No           49.574949      118.28022162236736   
2                     No           49.574949         81.699521264648   
3                     No           49.574949       199.4580743910713   
4                     No           49.574949      41.420153086217326   

                  Payment_Behaviour     Monthly_Balance Credit_Score  
0   High_spent_Small_value_payments  312.49408867943663         Good  
1    Low_spent_Large_value_payments  284.62916249607184         Good  
2   Low_spent_Medium_value_payments   331.2098628537912         Good  
3    Low_spent_Small_value_payments  223.45130972736786         Good  
4  High_spent_Medium_value_payments  341.48923103222177         Good  

[5 rows x 28 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      90015 non-null   object 
 4   Age                       100000 non-null  object 
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  object 
 8   Monthly_Inhand_Salary     84998 non-null   float64
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  int64  
 12  Num_of_Loan               100000 non-null  object 
 13  Type_of_Loan              88592 non-null   ob

Remover colunas que não serão usadas, não são importantes para o algoritmo

In [4]:
df.drop(columns=['ID', 'Customer_ID', 'Month', 'Name', 'SSN', 'Num_Bank_Accounts'], axis=1, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 22 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Age                       100000 non-null  object 
 1   Occupation                100000 non-null  object 
 2   Annual_Income             100000 non-null  object 
 3   Monthly_Inhand_Salary     84998 non-null   float64
 4   Num_Credit_Card           100000 non-null  int64  
 5   Interest_Rate             100000 non-null  int64  
 6   Num_of_Loan               100000 non-null  object 
 7   Type_of_Loan              88592 non-null   object 
 8   Delay_from_due_date       100000 non-null  int64  
 9   Num_of_Delayed_Payment    92998 non-null   object 
 10  Changed_Credit_Limit      100000 non-null  object 
 11  Num_Credit_Inquiries      98035 non-null   float64
 12  Credit_Mix                100000 non-null  object 
 13  Outstanding_Debt          100000 non-null  ob

Transformar a variável Age para quantitativa

In [5]:
df['Age'] = pd.to_numeric(df['Age'], errors='coerce')
df['Age'] = df['Age'].astype('Int64')

Estatística das variáveis quantitativas

In [6]:
df.describe()

Age  Monthly_Inhand_Salary  Num_Credit_Card  Interest_Rate  \
count     95061.0           84998.000000     100000.00000  100000.000000   
mean   110.934505            4194.170850         22.47443      72.466040   
std    689.407864            3183.686167        129.05741     466.422621   
min        -500.0             303.645417          0.00000       1.000000   
25%          24.0            1625.568229          4.00000       8.000000   
50%          33.0            3093.745000          5.00000      13.000000   
75%          42.0            5957.448333          7.00000      20.000000   
max        8698.0           15204.633333       1499.00000    5797.000000   

       Delay_from_due_date  Num_Credit_Inquiries  Credit_Utilization_Ratio  \
count        100000.000000          98035.000000             100000.000000   
mean             21.068780             27.754251                 32.285173   
std              14.860104            193.177339                  5.116875   
min              -5.000000              0.000000                 20.000000   
25%              10.000000              3.000000                 28.052567   
50%              18.000000              6.000000                 32.305784   
75%              28.000000              9.000000                 36.496663   
max              67.000000           2597.000000                 50.000000   

       Total_EMI_per_month  
count        100000.000000  
mean           1403.118217  
std            8306.041270  
min               0.000000  
25%              30.306660  
50%              69.249473  
75%             161.224249  
max           82331.000000

Algumas observações sobre as estatísticas acima:

* Age: temos que investigar, pois os valores mínimo e máximo estão bem estranhos

* Num_Bank_Accounts: o valor min e max com certeza está com problema, pois com certeza ninguém tem -1 conta em banco ou quase duas mil contas...

* Num_Credit_Card: O valor máximo de 1499 também está errado e precisa ser tratado.

* Interest_Rate: Valor máximo precisa ser tratado

* Delay_from_due_date: min = -5 precisa ser tratado

* Num_Credit_Inquiries: tratar valor máximo

Verificando se há valores ausentes

In [7]:
df.isna().sum()

Age                          4939
Occupation                      0
Annual_Income                   0
Monthly_Inhand_Salary       15002
Num_Credit_Card                 0
Interest_Rate                   0
Num_of_Loan                     0
Type_of_Loan                11408
Delay_from_due_date             0
Num_of_Delayed_Payment       7002
Changed_Credit_Limit            0
Num_Credit_Inquiries         1965
Credit_Mix                      0
Outstanding_Debt                0
Credit_Utilization_Ratio        0
Credit_History_Age           9030
Payment_of_Min_Amount           0
Total_EMI_per_month             0
Amount_invested_monthly      4479
Payment_Behaviour               0
Monthly_Balance              1200
Credit_Score                    0
dtype: int64

Como esses valores são uma pequena proporção, dado o tamanho do conjunto de dados, optei por excluí-los, menos os da coluna 'Monthly_Inhand_Salary', os quais vou preencher com a mediana.

In [8]:
# Preenchendo os valores ausentes da coluna 'Monthly_Inhand_Salary' com a Mediana
df['Monthly_Inhand_Salary'].fillna(df['Monthly_Inhand_Salary'].median(), inplace=True)

In [9]:
# Exluir os demais valores ausentes
df_sem_nulos = df.dropna()

In [10]:
df_sem_nulos.isna().sum()

Age                         0
Occupation                  0
Annual_Income               0
Monthly_Inhand_Salary       0
Num_Credit_Card             0
Interest_Rate               0
Num_of_Loan                 0
Type_of_Loan                0
Delay_from_due_date         0
Num_of_Delayed_Payment      0
Changed_Credit_Limit        0
Num_Credit_Inquiries        0
Credit_Mix                  0
Outstanding_Debt            0
Credit_Utilization_Ratio    0
Credit_History_Age          0
Payment_of_Min_Amount       0
Total_EMI_per_month         0
Amount_invested_monthly     0
Payment_Behaviour           0
Monthly_Balance             0
Credit_Score                0
dtype: int64

Verificando os valores da coluna Age

In [11]:
unique_values_age = df_sem_nulos['Age'].unique()
print(unique_values_age)

<IntegerArray>
[  23, -500,   28,   34,   54,   55,   33,   30,   24,   44,
 ...
 1953,  278, 1683, 8204, 7409, 5642, 8315, 8425, 2263, 1342]
Length: 1210, dtype: Int64


In [12]:
# Definir limites aceitáveis para a idade
idade_minima = 18
idade_maxima = 100

# Filtrar os valores fora do intervalo aceitável
valores_absurdos = df_sem_nulos[(df['Age'] < idade_minima) | (df_sem_nulos['Age'] > idade_maxima)]

# Contar quantos valores estão fora do intervalo aceitável
quantidade_valores_absurdos = valores_absurdos.shape[0]

print(f'Quantidade de valores absurdos na coluna "Age": {quantidade_valores_absurdos}')

Quantidade de valores absurdos na coluna "Age": 6051


C:\Users\marce\AppData\Local\Temp\ipykernel_16608\2182923767.py:6: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  valores_absurdos = df_sem_nulos[(df['Age'] < idade_minima) | (df_sem_nulos['Age'] > idade_maxima)]


In [13]:
# Vamos remover esses valores
df_clean = df_sem_nulos.drop(valores_absurdos.index)

In [14]:
unique_values_delay = df_sem_nulos['Delay_from_due_date'].sort_values(ascending=True).unique()
print(unique_values_delay)

[-5 -4 -3 -2 -1  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18
 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42
 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66
 67]


In [15]:
delay_min = 0
delay_max = 67

valores_fora = df_clean[(df['Delay_from_due_date'] < delay_min) | (df_clean['Delay_from_due_date'] > delay_max)]

quantidade_valores_fora = valores_fora.shape[0]

print(quantidade_valores_fora)

336


C:\Users\marce\AppData\Local\Temp\ipykernel_16608\3201568352.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  valores_fora = df_clean[(df['Delay_from_due_date'] < delay_min) | (df_clean['Delay_from_due_date'] > delay_max)]


In [16]:
df_clean = df_clean.drop(valores_fora.index)

In [17]:
# transformar coluna 'Num_of_Delayed_Payment' para quantitativa
df_clean['Num_of_Delayed_Payment'] = pd.to_numeric(df['Num_of_Delayed_Payment'], errors='coerce')
df_clean['Num_of_Delayed_Payment'].astype('Int64')

0           7
3           4
5           4
6        <NA>
9           1
         ... 
99994       6
99995       7
99996       7
99997       6
99999       6
Name: Num_of_Delayed_Payment, Length: 59513, dtype: Int64

In [18]:
df_clean['Num_of_Delayed_Payment'].isna().sum()

1748

In [19]:
df_clean.dropna(subset='Num_of_Delayed_Payment', inplace=True)
df_clean.isna().sum()

Age                         0
Occupation                  0
Annual_Income               0
Monthly_Inhand_Salary       0
Num_Credit_Card             0
Interest_Rate               0
Num_of_Loan                 0
Type_of_Loan                0
Delay_from_due_date         0
Num_of_Delayed_Payment      0
Changed_Credit_Limit        0
Num_Credit_Inquiries        0
Credit_Mix                  0
Outstanding_Debt            0
Credit_Utilization_Ratio    0
Credit_History_Age          0
Payment_of_Min_Amount       0
Total_EMI_per_month         0
Amount_invested_monthly     0
Payment_Behaviour           0
Monthly_Balance             0
Credit_Score                0
dtype: int64

In [20]:
unique_values_inquiries = df_clean['Num_Credit_Inquiries'].unique()
print(unique_values_inquiries)

[4.000e+00 2.000e+00 3.000e+00 5.000e+00 9.000e+00 8.000e+00 7.000e+00
 6.000e+00 0.000e+00 1.000e+00 1.000e+01 1.050e+03 1.100e+01 1.044e+03
 1.700e+01 1.936e+03 1.300e+01 5.680e+02 1.618e+03 5.250e+02 1.200e+01
 1.251e+03 1.400e+01 1.600e+01 1.500e+01 2.850e+02 1.241e+03 1.839e+03
 1.953e+03 1.265e+03 1.085e+03 1.194e+03 1.197e+03 9.960e+02 4.180e+02
 8.290e+02 1.760e+03 2.539e+03 5.460e+02 3.300e+01 6.020e+02 1.122e+03
 2.134e+03 1.151e+03 9.550e+02 7.940e+02 9.770e+02 2.158e+03 3.040e+02
 8.600e+01 1.347e+03 2.161e+03 1.529e+03 1.615e+03 2.187e+03 2.060e+02
 5.700e+01 1.970e+02 4.580e+02 1.114e+03 3.210e+02 2.572e+03 1.479e+03
 2.050e+02 1.890e+03 2.081e+03 9.460e+02 1.393e+03 1.244e+03 1.015e+03
 5.740e+02 3.640e+02 1.498e+03 1.562e+03 2.030e+02 2.297e+03 3.400e+01
 2.128e+03 2.207e+03 8.240e+02 5.360e+02 2.056e+03 5.280e+02 1.247e+03
 3.990e+02 3.930e+02 2.210e+02 2.328e+03 2.548e+03 2.330e+02 2.580e+03
 2.245e+03 1.921e+03 1.173e+03 1.010e+02 2.326e+03 1.278e+03 1.999e+03
 2.308

In [21]:
# Vamos converter esses valores para decimal padrão
df_clean['Num_Credit_Inquiries'] = pd.to_numeric(df_clean['Num_Credit_Inquiries'], errors='coerce')
print(df_clean['Num_Credit_Inquiries'].head())

0     4.0
3     4.0
5     4.0
9     2.0
10    2.0
Name: Num_Credit_Inquiries, dtype: float64


In [22]:
unique_values_utilization = df_clean['Num_Credit_Inquiries'].unique()
print(unique_values_utilization)

[4.000e+00 2.000e+00 3.000e+00 5.000e+00 9.000e+00 8.000e+00 7.000e+00
 6.000e+00 0.000e+00 1.000e+00 1.000e+01 1.050e+03 1.100e+01 1.044e+03
 1.700e+01 1.936e+03 1.300e+01 5.680e+02 1.618e+03 5.250e+02 1.200e+01
 1.251e+03 1.400e+01 1.600e+01 1.500e+01 2.850e+02 1.241e+03 1.839e+03
 1.953e+03 1.265e+03 1.085e+03 1.194e+03 1.197e+03 9.960e+02 4.180e+02
 8.290e+02 1.760e+03 2.539e+03 5.460e+02 3.300e+01 6.020e+02 1.122e+03
 2.134e+03 1.151e+03 9.550e+02 7.940e+02 9.770e+02 2.158e+03 3.040e+02
 8.600e+01 1.347e+03 2.161e+03 1.529e+03 1.615e+03 2.187e+03 2.060e+02
 5.700e+01 1.970e+02 4.580e+02 1.114e+03 3.210e+02 2.572e+03 1.479e+03
 2.050e+02 1.890e+03 2.081e+03 9.460e+02 1.393e+03 1.244e+03 1.015e+03
 5.740e+02 3.640e+02 1.498e+03 1.562e+03 2.030e+02 2.297e+03 3.400e+01
 2.128e+03 2.207e+03 8.240e+02 5.360e+02 2.056e+03 5.280e+02 1.247e+03
 3.990e+02 3.930e+02 2.210e+02 2.328e+03 2.548e+03 2.330e+02 2.580e+03
 2.245e+03 1.921e+03 1.173e+03 1.010e+02 2.326e+03 1.278e+03 1.999e+03
 2.308

In [23]:
df_clean['Credit_Utilization_Ratio'] = pd.to_numeric(df_clean['Credit_Utilization_Ratio'], errors='coerce')
print(df_clean['Credit_Utilization_Ratio'].head())

0     26.822620
3     31.377862
5     27.262259
9     38.550848
10    33.224951
Name: Credit_Utilization_Ratio, dtype: float64


### Escolha das colunas para os algoritmos de classificação

features: age, occupation, annual_income, delay_drom_due_date, num_of_delayed_payment, num_credit_inquiries, credit_utilization_rate

target: credit_score

In [24]:
colunas = ['Age', 'Occupation', 'Annual_Income', 'Delay_from_due_date', 'Num_of_Delayed_Payment', 'Num_Credit_Inquiries', 'Credit_Utilization_Ratio']

tipos_de_dados = df_clean[colunas].dtypes
print(tipos_de_dados)

Age                           Int64
Occupation                   object
Annual_Income                object
Delay_from_due_date           int64
Num_of_Delayed_Payment      float64
Num_Credit_Inquiries        float64
Credit_Utilization_Ratio    float64
dtype: object


Transformar coluna Annual_Income para numérica

In [25]:
df_clean['Annual_Income'] = pd.to_numeric(df_clean['Annual_Income'], errors='coerce')
df_clean['Annual_Income'].astype('float32')

0        19114.119141
3        19114.119141
5        19114.119141
9        34847.839844
10                NaN
             ...     
99994    39628.988281
99995    39628.988281
99996    39628.988281
99997    39628.988281
99999             NaN
Name: Annual_Income, Length: 57765, dtype: float32

In [26]:
# Eliminar os NaN que surgiram na coluna
df_clean.dropna(subset=['Annual_Income'], inplace=True)

In [27]:
df_clean['Occupation'].value_counts()

Occupation
_______          3775
Lawyer           3604
Mechanic         3481
Engineer         3409
Architect        3406
Scientist        3360
Teacher          3360
Doctor           3360
Developer        3343
Entrepreneur     3329
Accountant       3322
Media_Manager    3292
Journalist       3219
Writer           3172
Musician         3167
Manager          3158
Name: count, dtype: int64

Na coluna Occupation tem o valor '------', vamos exlcuir

In [28]:
df_clean['Occupation'].replace('_______', pd.NA, inplace=True)
df_clean.dropna(subset=['Occupation'], inplace=True)
print(df_clean['Occupation'].value_counts())

Occupation
Lawyer           3604
Mechanic         3481
Engineer         3409
Architect        3406
Scientist        3360
Teacher          3360
Doctor           3360
Developer        3343
Entrepreneur     3329
Accountant       3322
Media_Manager    3292
Journalist       3219
Writer           3172
Musician         3167
Manager          3158
Name: count, dtype: int64


Transformar a coluna Occupation em numérica usando o Frequency Encoding: técnica que transforma cada categoria em seu respectivo número de ocorrência na coluna, substituindo a categoria pelo número de vezes que ela aparece. 

In [29]:
# Calcular a contagem de ocorrência de cada vlaor na coluna
freq_ocupacao = df_clean['Occupation'].value_counts(normalize=True)

# Mapear os valores da coluna "Occupation" para as suas frequências
df_clean['Occupation_Frequency_Encoded'] = df_clean['Occupation'].map(freq_ocupacao)

# Exibindo as primeiras linhas após a codificação
print(df_clean[['Occupation', 'Occupation_Frequency_Encoded']].head())

   Occupation  Occupation_Frequency_Encoded
0   Scientist                      0.067224
3   Scientist                      0.067224
5   Scientist                      0.067224
9     Teacher                      0.067224
12    Teacher                      0.067224


Atribuir um valor numérico a cada categoria da coluna 'Credit_Score'

In [30]:
df_clean['Credit_Score'].value_counts()

Credit_Score
Standard    26426
Poor        15193
Good         8363
Name: count, dtype: int64

In [31]:
score_mapping = {'Standard': 2, 'Poor': 1, 'Good': 3}

df_clean['Credit_Score_Numeric'] = df_clean['Credit_Score'].map(score_mapping)

print(df_clean[['Credit_Score', 'Credit_Score_Numeric']].head())

   Credit_Score  Credit_Score_Numeric
0          Good                     3
3          Good                     3
5          Good                     3
9          Good                     3
12         Good                     3


Importar as bibliotecas

In [32]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.neighbors import KNeighborsClassifier

Escolhendo as variáveis explicativas (features) e a alvo (target)

In [33]:
features = ['Age', 'Occupation_Frequency_Encoded', 'Annual_Income', 'Delay_from_due_date', 'Num_of_Delayed_Payment', 'Num_Credit_Inquiries', 'Credit_Utilization_Ratio']
target = 'Credit_Score_Numeric'

X = df_clean[features]
y = df_clean[target]

Dividir em dados de treino e teste

In [34]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=42)

### 1. KNN

Inicializando e treinando o modelo 

In [35]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

KNeighborsClassifier()

Fazendo previsões com o modelo

In [36]:
knn_prev = knn.predict(X_test)

Avaliando o modelo

In [37]:
print(f'Relatório:\n {classification_report(y_test, knn_prev)}')

Relatório:
               precision    recall  f1-score   support

           1       0.67      0.73      0.70      2986
           2       0.74      0.75      0.75      5350
           3       0.66      0.54      0.59      1661

    accuracy                           0.71      9997
   macro avg       0.69      0.67      0.68      9997
weighted avg       0.71      0.71      0.71      9997



### 2. RANDOM FOREST

Inicializar e treinar o modelo

In [38]:
rnd_forest = RandomForestClassifier(criterion='entropy', max_depth=25)
rnd_forest.fit(X_train, y_train)

RandomForestClassifier(criterion='entropy', max_depth=25)

Previsões

In [39]:
prev_rnd_forest = rnd_forest.predict(X_test)

Avaliação do modelo

In [40]:
print(f'Relatório:\n {classification_report(y_test, prev_rnd_forest)}')

Relatório:
               precision    recall  f1-score   support

           1       0.72      0.71      0.71      2986
           2       0.73      0.77      0.75      5350
           3       0.66      0.54      0.59      1661

    accuracy                           0.72      9997
   macro avg       0.70      0.67      0.69      9997
weighted avg       0.71      0.72      0.71      9997



### 3. XGBoost

Para poder usar o XGBoost os rótulos de y tem que estar no formato esperado, começar por 0

In [41]:
import numpy as np

y = np.array(y) - 1


In [42]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=42)

In [43]:
unique_classes = np.unique(y)
print(f'Classes corrigidas {unique_classes}')

Classes corrigidas [0 1 2]


Inicializar e treinar o modelo

In [44]:
xg = XGBClassifier()
xg.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, objective='multi:softprob', ...)

Previsões

In [45]:
xg_prev = xg.predict(X_test)

Avaliação do modelo

In [46]:
print(f'Relatório: {classification_report(y_test, xg_prev)}')

Relatório:               precision    recall  f1-score   support

           0       0.67      0.66      0.66      2986
           1       0.69      0.75      0.72      5350
           2       0.59      0.47      0.52      1661

    accuracy                           0.67      9997
   macro avg       0.65      0.62      0.63      9997
weighted avg       0.67      0.67      0.67      9997



### CONCLUSÃO

Usando os parâmetros padrões dos algoritmos, o que melhor resultado teve na classficação foi o KNN, seguido do Random Forest.